addons/l10n_it_edi/models/account_move.py

```diff
@@ -0,0 +1,1212 @@
+        tax_lines = self._l10n_it_edi_get_tax_values(tax_details)
+
+        # Reduce downpayment views to a single recordset
+        downpayment_moves = [l.get('downpayment_moves', self.env['account.move']) for l in invoice_lines]
+        downpayment_moves = self.browse(move.id for moves in downpayment_moves for move in moves)
+
+        return {
+            'record': self,
+            'company': company,
+            'partner': partner,
+            'sender': sender,
+            'buyer': buyer,
+            'seller': seller,
+            'representative': company.l10n_it_tax_representative_partner_id,
+            'sender_info': sender_info_values,
+            'buyer_info': buyer_info_values,
+            'seller_info': seller_info_values,
+            'representative_info': representative_info_values,
+            'currency': self.currency_id or self.company_currency_id if not convert_to_euros else self.env.ref('base.EUR'),
+            'document_total': document_total,
+            'regime_fiscale': company.l10n_it_tax_system if not is_self_invoice else 'RF18',
+            'is_self_invoice': is_self_invoice,
+            'partner_bank': self.partner_bank_id,
+            'formato_trasmissione': formato_trasmissione,
+            'document_type': document_type,
+            'tax_details': tax_details,
+            'downpayment_moves': downpayment_moves,
+            'rc_refund': reverse_charge_refund,
+            'invoice_lines': invoice_lines,
+            'tax_lines': tax_lines,
+            'conversion_rate': conversion_rate,
+            'balance_multiplicator': -1 if self.is_inbound() else 1,
+            'abs': abs,
+            'pdf_name': pdf_values['name'] if pdf_values else False,
+            'pdf': b64encode(pdf_values['raw']).decode() if pdf_values else False,
+        }
+
+    def _l10n_it_edi_services_or_goods(self):
+        """
+            Services and goods have different tax grids when VAT is Reverse Charged, and they can't
+            be mixed in the same invoice, because the TipoDocumento depends on which which kind
+            of product is bought and it's unambiguous.
+        """
+        self.ensure_one()
+        scopes = []
+        for line in self.invoice_line_ids.filtered(lambda l: l.display_type not in ('line_note', 'line_section')):
+            tax_ids_with_tax_scope = line.tax_ids.filtered(lambda x: x.tax_scope)
+            if tax_ids_with_tax_scope:
+                scopes += tax_ids_with_tax_scope.mapped('tax_scope')
+            else:
+                scopes.append(line.product_id and line.product_id.type or 'consu')
+
+        if set(scopes) == {'consu', 'service'}:
+            return "both"
+        return scopes and scopes.pop()
+
+    def _compute_l10n_it_edi_is_self_invoice(self):
+        for move in self:
+            move.l10n_it_edi_is_self_invoice = move._l10n_it_edi_is_self_invoice()
+
+    def _l10n_it_edi_is_self_invoice(self):
```

Comment:
remove this method and use only the computed field instead?

Reply:
I assume this is intentional by @lordkrandel to avoid the overhead caused by the ORM's caching and recomputation features.